### ISSUE TITLE
Lasagne/examples/recurrent.py

### ISSUE THREAD

**Issue by George19890617** [Lasagne/Lasagne#733]
I am not clear about what is doing in this example?

1. as you said to "output the sum of two numbers in a sequence ..." , but in the code it seem to sum the product, "y[n] = np.sum(X[n, :, 0]*X[n, :, 1])"

Recurrent network example.  Trains a bidirectional vanilla RNN to output the
sum of two numbers in a sequence of random numbers sampled uniformly from
[0, 1] based on a separate marker sequence.

2. I am real confused by the notes here? 
Generate a batch of sequences for the "add" task, e.g. the target for the
    following
    ``| 0.5 | 0.7 | 0.3 | 0.1 | 0.2 | ... | 0.5 | 0.9 | ... | 0.8 | 0.2 |
      |  0  |  0  |  1  |  0  |  0  |     |  0  |  1  |     |  0  |  0  |``
    would be 0.3 + .9 = 1.2.  This task was proposed in [1]_ and explored in
    e.g. [2]_.

**Comment by AlfioEmanueleFresta** [comment]
@George19890617 in the example, you sum numbers in the first row only where you have a 1 in the second row.

    X[..., 0] = | 0.5 | 0.7 | 0.3 | 0.1 | 0.2 | ... | 0.5 | 0.9 | ... | 0.8 | 0.2 |
    X[..., 1] = |  0  |  0  |  1  |  0  |  0  | ... |  0  |  1  | ... |  0  |  0  |

Therefore, the output Y for the sequence above is equivalent to:

    Y[...] = 0.5 * 0 + 0.7 * 0 + 0.3 * 1 + ... + 0.2 * 0

Which should explain the line from the code:

```python
y[...] = np.sum(X[..., 0] * X[..., 1])
```

**Comment by f0k** [comment]
> but in the code it seem to sum the product

`X[n, :, 0]` is the first feature (the random numbers) and `X[n, :, 1]` is the second feature (the mask values). So it does an elementwise multiplication of the sequence of random numbers with the sequence of mask values (which will only have two values surviving) and then computes the sum of the resulting sequence, which adds up the two surviving numbers.

Closing this because I assume you understood the documentation correctly, but just weren't sure whether the code matched your understanding of the documentation. Let us know if the documentation should be improved in any way.

**Comment by George19890617** [comment]
thanks for the answering, So if the program is try to let the lstm understand the add task, but why in the gen_data, center the data, say:
# Center the inputs and outputs
        X -= X.reshape(-1, 2).mean(axis=0)
        y -= y.mean()

Because I think after the centering, the y is not anymore the sum of  X.


在 2016年8月10日，上午7:40，Alfio Emanuele <notifications@github.com<mailto:notifications@github.com>> 写道：


@George19890617<https://github.com/George19890617> in the example, you sum numbers in the first row only where you have a 1 in the second row.

X[..., 0] = | 0.5 | 0.7 | 0.3 | 0.1 | 0.2 | ... | 0.5 | 0.9 | ... | 0.8 | 0.2 |
X[..., 1] = |  0  |  0  |  1  |  0  |  0  | ... |  0  |  1  | ... |  0  |  0  |


Therefore, the output Y for the sequence above is equivalent to:

Y[...] = 0.5 * 0 + 0.7 * 0 + 0.3 * 1 + ... + 0.2 * 0


Which should explain the line from the code:

y[...] = np.sum(X[..., 0] * X[..., 1])

**Comment by f0k** [comment]
> Because I think after the centering, the y is not anymore the sum of X.

True. In general, this wouldn't be a problem -- if all the inputs and outputs are offset by a constant value, the network can counter this by its biases (and it makes learning easier if the inputs and outputs have zero mean).
But there's a catch: That code will use a different offset for every batch (https://github.com/Lasagne/Lasagne/blob/master/examples/recurrent.py#L163)! I don't see how this would work -- unless I'm missing something, there will be a different relation from input to output in every batch. @craffel, what's up with that?

Reopening because this seems to need attention.

**Comment by craffel** [comment]
> That code will use a different offset for every batch 

This shouldn't be an issue - the network should be able to learn the normalization too (right?).  Maybe I'm remembering incorrectly, but doesn't the network eventually achieve perfect accuracy?  FWIW, I am doing this because it's done in "On the importance of initialization and momentum in deep learning":

"Finally, we found centering (mean subtraction) of both the inputs and the outputs to be important to reliably solve all of the training problems."

**Comment by f0k** [comment]
> This shouldn't be an issue - the network should be able to learn the normalization too (right?).

Well, the problem is that the normalization is different for every batch. For the second input column it could theoretically recognize the offset upon seeing the first time step, since the original data can only have been 0 and 1. For the first input column, the offset will be about 0.5, but always a little different. For the target vector, the offset will be about 1.0, since it's the sum of two random variables that each have an expected value of 0.5. Again, it will be a little different for each batch.

Demonstration (X.reshape(-1, 2).mean(axis=0), y.mean()):
```
(array([ 0.47155581,  0.03636364]), 1.0508654847494268)
(array([ 0.47281006,  0.03636364]), 1.0372341429506233)
(array([ 0.47582821,  0.03636364]), 1.0138743277560949)
(array([ 0.46950389,  0.03636364]), 0.94383995306177593)
(array([ 0.46911562,  0.03636364]), 0.93098336968475837)
(array([ 0.48026337,  0.03636364]), 1.0073445350326233)
(array([ 0.47312641,  0.03636364]), 0.95579423438279687)
(array([ 0.47407438,  0.03636364]), 1.0361235583483566)
(array([ 0.47758101,  0.03636364]), 1.0695646416473297)
(array([ 0.46817421,  0.03636364]), 0.91693917714132811)
```

> Maybe I'm remembering incorrectly, but doesn't the network eventually achieve perfect accuracy?

If so, I'm surprised how :) Am I missing something?

> "Finally, we found centering (mean subtraction) of both the inputs and the outputs to be important to reliably solve all of the training problems."

Yes, but this should be done across the dataset, not per batch.

Would you be okay if I changed the example to use the expected constant offsets instead of the batchwise mean?

**Comment by craffel** [comment]
>  For the target vector, the offset will be about 1.0, since it's the sum of two random variables that each have an expected value of 0.5. Again, it will be a little different for each batch.

For sure, but as the network sees more batches, it should be able to learn to predict the expected value over this per-batch normalization noise.

> If so, I'm surprised how :) Am I missing something?

Accuracy is ([historically](http://deeplearning.cs.cmu.edu/pdfs/Hochreiter97_lstm.pdf)) defined as predicting the target within .04 absolute.  I think it should be able to achieve this (or at least, extremely close) as-is.  Though, looking at the example, I'm not actually computing that accuracy metric, so I guess that comment was out-of-place.

> Would you be okay if I changed the example to use the expected constant offsets instead of the batchwise mean?

Yes, fine with that.

**Comment by f0k** [comment]
> For sure, but as the network sees more batches, it should be able to learn to predict the expected value over this per-batch normalization noise.

Yes, but it will not be able to predict the value that happened to be used for the validation batch.

> Accuracy is (historically) defined as predicting the target within .04 absolute.

Ah, I see. But the target noise seems to be larger than that.

Just to make sure, I've ran the original example for 200 epochs, ending with:
```
Epoch 190 validation cost = 0.0173958614469
Epoch 191 validation cost = 0.00847106706351
Epoch 192 validation cost = 0.00947748869658
Epoch 193 validation cost = 0.00830858014524
Epoch 194 validation cost = 0.0105759454891
Epoch 195 validation cost = 0.00689136143774
Epoch 196 validation cost = 0.00876523274928
Epoch 197 validation cost = 0.0111606745049
Epoch 198 validation cost = 0.0124800130725
Epoch 199 validation cost = 0.0111360838637
```
It lingers around 0.01 and does not improve further -- results mainly depend on how well the current biases fit the offset used in the validation batch, I guess. Note that this is mean squared error, so the absolute error would be around 0.1.

With constant offsets:
```python
X -= (0.5, 2.0 / n_batch / max_length)
y -= 1.0
```
I get:
```
Epoch 190 validation cost = 0.00439414521679
Epoch 191 validation cost = 0.00428196648136
Epoch 192 validation cost = 0.00457407766953
Epoch 193 validation cost = 0.00442864000797
Epoch 194 validation cost = 0.00439688470215
Epoch 195 validation cost = 0.00430428842083
Epoch 196 validation cost = 0.00422209268436
Epoch 197 validation cost = 0.00445842323825
Epoch 198 validation cost = 0.00416272319853
Epoch 199 validation cost = 0.00411068694666
```
Much better and still not overfitting, but not solved according to the 0.04 absolute error metric.

Maybe we should implement that metric as well and adjust the hyperparameters.

**Comment by f0k** [comment]
> Much better and still not overfitting, but not solved according to the 0.04 absolute error metric.

> Maybe we should implement that metric as well and adjust the hyperparameters.

Hmm, the example does not exactly follow any of these existing variants:
* [Hochreiter et al 1997](http://deeplearning.cs.cmu.edu/pdfs/Hochreiter97_lstm.pdf), Sect. 5.4
* [Martens et al 2011](http://www.cs.utoronto.ca/~ilya/pubs/2011/HF-RNN_supp.pdf), Sect. 2.1
* [Sutskever et al 2013](www.cs.toronto.edu/~fritz/absps/momentum.pdf), Sect. A.4

Hochreiter used uniform samples from (-1, 1), marked the first and last time step with -1, sampled the second marked item within (0, T/2 - 1), and set the target to 0.5 + (x1 + x2) / 4.
Martens used uniform samples from (0, 1) and marked items as we do, but sampled the second marked item within (T/10, T/2), and set the target to (x1 + x2) / 2.
Sutskever is not very clear, they both say that the target is (x1 + x2) and that they follow Martens.
Both Hochreiter and Martens used the < 0.04 metric, which are compatible (the different input scaling is countered by the different target scaling).

I'm inclined to follow Martens, but set the target to (x1 + x2) and evaluate with < 0.08. If we set the target to (x1 + x2) / 2, people will be confused again.

**Comment by craffel** [comment]
> Hmm, the example does not exactly follow any of these existing variants:

Yes; the point of this example is not to show state-of-the-art results on an exactly-specified standard task, but instead to show basic usage of Lasagne's RNN functionality.  It differs from the MNIST example in this respect.  As you've surmised, there's not a ton of consensus about how this task should be implemented (partially because it's a toy task, and achieving high accuracy for a short sequence is much less interesting than achieving better-than-guessing on a long sequence).  This is one reason I am not reporting any notion of accuracy.  If you think it'd be better to make the example more closely approach usage according to some specific paper(s), that's fine, but I don't personally think it's particularly beneficial.
